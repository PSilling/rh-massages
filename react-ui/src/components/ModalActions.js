// react imports
import React, { Component } from 'react';

// util imports
import _t from '../utils/Translations.js';

/**
 * Custom component rendering action buttons for modal footers.
 */
class ModalActions extends Component {

  render() {
    return(
      <div className="pull-right">
        <button type="button" className="btn btn-primary" onClick={this.props.onProceed}
          style={{ 'marginRight': '5px' }} autoFocus={this.props.autoFocus}>
          {this.props.primaryLabel}
        </button>
        <button type="button" className="btn btn-default"
          onClick={this.props.onClose}>
          { _t.translate('Dismiss') }
        </button>
      </div>
    );
  }
}

export default ModalActions
