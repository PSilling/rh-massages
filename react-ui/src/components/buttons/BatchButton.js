// react imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * A default styled button component with a given label. Used primarily for
 * batch operation buttons.
 */
class BatchButton extends Component {

  render() {
    return (
      <span>
        <button type="button" className="btn btn-default" onClick={this.props.onClick}
          disabled={this.props.disabled}>
          {this.props.label}
        </button>
      </span>
    )
  }
}

BatchButton.propTypes = {
  /** function to be called on button click */
  onClick: PropTypes.func.isRequired,
   /** button label */
  label: PropTypes.string,
  /** whether the button should be disabled */
  disabled: PropTypes.bool
};

BatchButton.defaultProps = {
  disabled: false
};

export default BatchButton
