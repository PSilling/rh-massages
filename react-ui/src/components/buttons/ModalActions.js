// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Footer buttons for Modal dialogs. Children components can be given to add more
 * case specific elements.
 */
const ModalActions = function ModalActions(props) {
  return (
    <div className="pull-right">
      {props.children}
      {props.primaryLabel === "none" ? (
        ""
      ) : (
        <button
          type="button"
          className="btn btn-primary"
          onClick={props.onProceed}
          title={props.title}
          style={{ marginRight: "5px" }}
          autoFocus={props.autoFocus}
          disabled={props.disabled}
        >
          {props.primaryLabel}
        </button>
      )}
      <button type="button" className="btn btn-default" onClick={props.onClose}>
        {_t.translate("Dismiss")}
      </button>
    </div>
  );
};

ModalActions.propTypes = {
  /** callback function triggered on close button click */
  onClose: PropTypes.func.isRequired,
  /** callback function triggered on primary button click */
  onProceed: PropTypes.func.isRequired,
  /** whether the primary button should be automatically focused */
  autoFocus: PropTypes.bool,
  /** extra elements to be provided before the two buttons */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** primary button label to be displayed */
  primaryLabel: PropTypes.string,
  /** primary button title */
  title: PropTypes.string
};

ModalActions.defaultProps = {
  autoFocus: false,
  children: null,
  disabled: false,
  primaryLabel: _t.translate("Proceed"),
  title: ""
};

export default ModalActions;
