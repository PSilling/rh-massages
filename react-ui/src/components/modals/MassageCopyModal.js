// react imports
import React, { Component } from 'react';
import ReactDOM from 'react-dom';

// component imports
import BatchButton from '../buttons/BatchButton';
import ModalActions from '../buttons/ModalActions';

// module imports
import {ModalContainer, ModalDialog} from 'react-modal-dialog';
import moment from 'moment';

// util imports
import _t from '../../util/Translations';
import Util from '../../util/Util';

class MassagaCopyModal extends Component {

  state = {count: 1, step: 7}

  changeCount = (event) => {
    if (Util.isEmpty(event.target.value) || parseInt(event.target.value, 10) < 1
        || parseInt(event.target.value, 10) > 54) {
      return;
    }
    this.setState({count: event.target.value});
  }

  changeStep = (event) => {
    if (Util.isEmpty(event.target.value) || parseInt(event.target.value, 10) < 1
        || parseInt(event.target.value, 10) > 365) {
      return;
    }
    this.setState({step: event.target.value});
  }

  /**
   * Handles the post request.
   */
  addMassages = () => {
    var postArray = [];

    for (var i = 1; i <= parseInt(this.state.count, 10); i++) {
      for (var j = 0; j < this.props.massages.length; j++) {
        postArray.push({
          date: moment(this.props.massages[j].date).add(this.state.step * i, 'days').toDate(),
          ending: moment(this.props.massages[j].ending).add(this.state.step * i, 'days').toDate(),
          masseuse: this.props.massages[j].masseuse,
          client: null,
          contact: null,
          facility: {id: this.props.massages[j].facility.id}
        });
      }
    }

    Util.post(Util.MASSAGES_URL, postArray, () => {
      this.props.getCallback();
      this.props.onToggle(true);
    });
  }

  handleModalKeyPress = (event) => {
    if (event.charCode === 13 && document.activeElement === ReactDOM.findDOMNode(this.modalDialog)) {
      this.addMassages();
    }
  }

  handleInputKeyPress = (event) => {
    if (event.key === 'Enter') {
      this.addMassages();
    }
  }

  moveCursorToEnd = (event) => {
    var value = event.target.value;
    event.target.value = '';
    event.target.value = value;
  }

  render() {
    return(
      <span style={{ 'marginRight': '5px' }}>
        <BatchButton onSubmit={() => this.props.onToggle(false)} disabled={this.props.disabled}
          label={ _t.translate('Copy selected') } />

        {this.props.active ?
          <ModalContainer onClose={() => this.props.onToggle(false)}>
            <ModalDialog onClose={() => this.props.onToggle(false)} width="50%" style={{ 'outline': 'none' }}
              tabIndex="1" onKeyPress={this.handleModalKeyPress}
              ref={(dialog) => {
                this.modalDialog = dialog;
              }}>
              <h2>
                { _t.translate('Copy Massages') }
              </h2>
              <hr />
              <form>
                <div className="form-group col-md-12">
                  <label>{ _t.translate('Number of repetitions') }</label>
                  <div className="row">
                    <div className="col-md-4">
                      <input value={this.state.count} onChange={this.changeCount}
                        className="form-control" onKeyPress={this.handleInputKeyPress}
                        autoFocus onFocus={this.moveCursorToEnd} type="number" min="1"
                        max="54" placeholder={ _t.translate('Number of repetitions') }
                      />
                    </div>
                  </div>
                </div>
                <div className="form-group col-md-12">
                  <label>{ _t.translate('Day step per repetition') }</label>
                  <div className="row">
                    <div className="col-md-4">
                      <input value={this.state.step} onChange={this.changeStep}
                        className="form-control" onKeyPress={this.handleInputKeyPress} onFocus={this.moveCursorToEnd}
                        type="number" min="1" max="365" placeholder={ _t.translate('Day step per repetition') }
                      />
                    </div>
                  </div>
                </div>
              </form>
              <ModalActions
                primaryLabel={ _t.translate('Copy') }
                onProceed={this.addMassages}
                onClose={() => this.props.onToggle(false)}
                autoFocus={false}
              />
            </ModalDialog>
          </ModalContainer> : ''
        }
      </span>
    )
  }
}

export default MassagaCopyModal
