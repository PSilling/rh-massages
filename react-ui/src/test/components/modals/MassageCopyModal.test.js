// react imports
import React from 'react';
import TestRenderer from 'react-test-renderer';

// test imports
import BatchButton from '../../../components/buttons/BatchButton';
import MassageCopyModal from '../../../components/modals/MassageCopyModal';
import ModalActions from '../../../components/buttons/ModalActions';

// test mocks
jest.mock('../../../util/Util');

afterEach(() => {
  jest.resetAllMocks();
});

test('renders inside content with correct props', () => {
  const testGetFunction = jest.fn(),
        testToggleFunction = jest.fn(),
        testMassages = [{
          id: 1,
          date: new Date(0),
          ending: new Date(1000),
          client: null,
          facility: { id: 1, name: "test" }
        }],
        testRenderer = TestRenderer.create(
          <MassageCopyModal active massages={testMassages} getCallback={testGetFunction}
            onToggle={testToggleFunction} withPortal={false} />
        ),
        testInstance = testRenderer.root;

  let button = testInstance.findByType(BatchButton),
      actions = testInstance.findByType(ModalActions),
      treeJSON = testRenderer.toJSON();

  expect(testToggleFunction).not.toHaveBeenCalled();

  button.props.onClick();
  actions.props.onClose();

  expect(testToggleFunction).toHaveBeenCalledTimes(2);
  expect(testToggleFunction).toHaveBeenLastCalledWith(false);
  expect(testGetFunction).not.toHaveBeenCalled();

  actions.props.onProceed();

  expect(testToggleFunction).toHaveBeenCalledTimes(3);
  expect(testToggleFunction).toHaveBeenLastCalledWith(true);
  expect(testGetFunction).toHaveBeenCalledTimes(1);
  expect(button.props.disabled).toBe(false);
  expect(treeJSON).toMatchSnapshot();
});
