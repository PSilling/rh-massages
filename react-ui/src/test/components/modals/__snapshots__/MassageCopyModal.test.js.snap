// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders inside content with correct props 1`] = `
<span
  style={
    Object {
      "marginRight": "5px",
    }
  }
>
  <span>
    <button
      className="btn btn-default"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Kopírovat vybrané
    </button>
  </span>
  <div>
    <h2>
      Zkopírovat masáže
    </h2>
    <hr />
    <div
      className="form-group col-md-12"
    >
      <label
        htmlFor="valueInput"
      >
        Počet opakování
      </label>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <input
            autoFocus={true}
            className="form-control"
            id="valueInput"
            max="54"
            min="1"
            onChange={[Function]}
            onFocus={undefined}
            onKeyPress={[Function]}
            placeholder="Počet opakování"
            type="number"
            value={1}
          />
        </div>
      </div>
    </div>
    <div
      className="form-group col-md-12"
    >
      <label
        htmlFor="stepInput"
      >
        Denní skok na opakování
      </label>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <input
            className="form-control"
            id="stepInput"
            max="365"
            min="1"
            onChange={[Function]}
            onFocus={undefined}
            onKeyPress={[Function]}
            placeholder="Denní skok na opakování"
            type="number"
            value={7}
          />
        </div>
      </div>
    </div>
    <div
      className="pull-right"
    >
      <button
        autoFocus={false}
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginRight": "5px",
          }
        }
        title=""
        type="button"
      >
        Zkopírovat
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="button"
      >
        Zrušit
      </button>
    </div>
  </div>
</span>
`;
